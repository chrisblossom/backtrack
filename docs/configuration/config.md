# config

`config` is used to modify a configuration set by a package/preset. See also: [config-manager.md](../config-manager.md).

```js
const backtrackConfig = {
    // ...
    config: [
        /**
         * Automatic config merge
         *
         * Performs a deep merge on the two objects.
         * Will append array items and keep existing object keys.
         */
        {
            '@backtrack/preset-style' /* config's namespace */: {
                rules: {
                    strict: 'off',
                },
            },
        },

        /**
         * Manual config merge
         *
         * Used when automatic merging does not merge as expected
         */
        {
            '@backtrack/preset-style' /* config's namespace */: (config) => {
                return {
                    ...config,
                    rules: {
                        ...config.rules,
                        strict: 'off',
                    },
                };
            },
        },
    ],
    // ...
};
```

```typescript
type Config = Array<{
    [key: string]: any;
}>;
```
