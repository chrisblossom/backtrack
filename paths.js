'use strict';

const paths = require('./dist/config/paths');

module.exports = paths;
