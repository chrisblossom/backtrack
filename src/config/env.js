/* @flow */

const env = {
    productionRelease: process.env.PRODUCTION_RELEASE !== 'false',

    development: process.env.NODE_ENV
        ? process.env.NODE_ENV === 'development'
        : true,

    /**
     * HMR enabled by default
     *
     * Always disable HMR in production-like environments
     */
    hmr:
        process.env.RUN_MODE === 'dev' && process.env.HMR !== 'false'
            ? process.env.NODE_ENV !== 'production'
            : false,
};

module.exports = Object.freeze(env);
