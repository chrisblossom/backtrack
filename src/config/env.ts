const productionRelease = process.env.PRODUCTION_RELEASE !== 'false';

const development = process.env.NODE_ENV
	? process.env.NODE_ENV === 'development'
	: true;

/**
 * HMR enabled by default
 *
 * Always disable HMR in production-like environments
 */
const hmr =
	process.env.RUN_MODE === 'dev' && process.env.HMR !== 'false'
		? process.env.NODE_ENV !== 'production'
		: false;

export { productionRelease, development, hmr };
