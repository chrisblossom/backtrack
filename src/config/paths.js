/* @flow */

import { realpathSync } from 'fs';
import path from 'path';

/**
 * Make sure any symlinks in the project folder are resolved:
 * https://github.com/facebookincubator/create-react-app/issues/637
 */
const rootPath = realpathSync(process.cwd());

const paths = {
    rootPath,
    buildPath: path.resolve(rootPath, 'dist'),
    sourcePath: path.resolve(rootPath, 'src'),
};

module.exports = Object.freeze(paths);
