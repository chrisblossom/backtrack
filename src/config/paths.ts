import { realpathSync } from 'fs';
import path from 'path';

/**
 * Make sure any symlinks in the project folder are resolved:
 * https://github.com/facebookincubator/create-react-app/issues/637
 */
const rootPath = realpathSync(process.cwd());
const buildPath = path.resolve(rootPath, 'dist');
const sourcePath = path.resolve(rootPath, 'src');

export { rootPath, buildPath, sourcePath };
