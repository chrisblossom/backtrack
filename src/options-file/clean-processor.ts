import { NormalizedClean } from '../types';

/**
 * Patterns are only allowed inside the buildPath.
 */
const base = {
	del: [],
	makeDirs: [],
	copy: [],
};

interface Args {
	value: NormalizedClean[];
	current?: NormalizedClean;
}

function cleanProcessor({ value, current = base }: Args) {
	return value.reduce(
		(acc, arg) => {
			arg.del.forEach((pattern) => {
				/**
				 * remove duplicates
				 */
				if (acc.del.includes(pattern) === false) {
					acc.del.push(pattern);
				}
			});

			arg.makeDirs.forEach((dir) => {
				/**
				 * remove duplicates
				 */
				if (acc.makeDirs.includes(dir) === false) {
					acc.makeDirs.push(dir);
				}
			});

			arg.copy.forEach((copy) => {
				/**
				 * remove duplicates
				 */
				const isDuplicate = !!acc.copy.find((copyArg) => {
					return (
						copyArg.src === copy.src && copyArg.dest === copy.dest
					);
				});

				if (isDuplicate === false) {
					acc.copy.push(copy);
				}
			});

			return acc;
		},
		{
			del: [...current.del],
			makeDirs: [...current.makeDirs],
			copy: [...current.copy],
		},
	);
}

export { cleanProcessor };
