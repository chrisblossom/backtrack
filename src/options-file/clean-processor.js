/* @flow */

import path from 'path';
import { buildPath } from '../config/paths';

import type { NormalizedClean } from '../types.js';

/**
 * Patterns are only allowed inside the buildPath.
 */
const base = {
    del: [],
    makeDirs: [],
};

// eslint-disable-next-line flowtype/require-exact-type
type Args = {
    value: $ReadOnlyArray<NormalizedClean>,
    current?: NormalizedClean,
};

function cleanProcessor({ value, current = base }: Args) {
    return value.reduce(
        (acc, arg) => {
            arg.del.forEach((pattern) => {
                /**
                 * remove duplicates
                 */
                if (acc.del.includes(pattern) === false) {
                    acc.del.push(pattern);
                }
            });

            arg.makeDirs.forEach((dir) => {
                const resolvedDir = path.resolve(buildPath, dir);

                /**
                 * remove duplicates
                 */
                if (acc.makeDirs.includes(resolvedDir) === false) {
                    acc.makeDirs.push(resolvedDir);
                }
            });

            return acc;
        },
        {
            del: [...current.del],
            makeDirs: [...current.makeDirs],
        },
    );
}

export { cleanProcessor };
