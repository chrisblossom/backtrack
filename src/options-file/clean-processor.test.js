/* @flow */

import path from 'path';
import { buildPath } from '../config/paths';

const cleanProcessor = (args) =>
    require('./clean-processor').cleanProcessor(args);

describe('cleanProcessor', () => {
    test('merges args together and completes resolves path', () => {
        const current = {
            del: ['first', 'second'],
            makeDirs: [
                path.resolve(buildPath, 'first'),
                path.resolve(buildPath, 'second'),
            ],
        };

        const value = [
            {
                del: ['third'],
                makeDirs: ['third'],
            },
        ];

        const result = cleanProcessor({ current, value });

        expect(result).toMatchSnapshot();
    });

    test('removes duplicates', () => {
        const value = [
            {
                del: ['first'],
                makeDirs: ['first'],
            },
            {
                del: ['first'],
                makeDirs: ['first'],
            },
        ];

        const result = cleanProcessor({ value });

        expect(result).toMatchSnapshot();
    });
});
