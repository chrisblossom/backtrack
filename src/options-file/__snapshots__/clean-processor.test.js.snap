// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cleanProcessor merges args together and completes resolves path 1`] = `
Object {
  "del": Array [
    "first",
    "second",
    "third",
  ],
  "makeDirs": Array [
    "<PROJECT_ROOT>/dist/first",
    "<PROJECT_ROOT>/dist/second",
    "<PROJECT_ROOT>/dist/third",
  ],
}
`;

exports[`cleanProcessor removes duplicates 1`] = `
Object {
  "del": Array [
    "first",
  ],
  "makeDirs": Array [
    "<PROJECT_ROOT>/dist/first",
  ],
}
`;
